model/code2seq.py

```diff
@@ -85,102 +60,38 @@ def _calculate_loss(self, logits: torch.Tensor, labels: torch.Tensor) -> torch.T
         return loss
 
     def _general_epoch_end(self, outputs: List[Dict], loss_key: str, group: str) -> Dict:
-        logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
-        logs.update(
-            SubtokenStatistic.union_statistics([out["subtoken_statistic"] for out in outputs]).calculate_metrics(group)
-        )
+        with torch.no_grad():
+            logs = {f"{group}/loss": torch.stack([out[loss_key] for out in outputs]).mean()}
+            logs.update(
+                SubtokenStatistic.union_statistics([out["statistic"] for out in outputs]).calculate_metrics(group)
+            )
         progress_bar = {k: v for k, v in logs.items() if k in [f"{group}/loss", f"{group}/f1"]}
         return {f"{group}_loss": logs[f"{group}/loss"], "log": logs, "progress_bar": progress_bar}
 
-    def transfer_batch_to_device(self, batch: PathContextBatch, device: torch.device) -> PathContextBatch:
-        # Dict str -> torch.Tensor [seq length; batch size * n_context]
-        for k in batch.context:
-            batch.context[k] = batch.context[k].to(self.device)
-        # [seq length; batch size]
-        batch.labels = batch.labels.to(self.device)
-        return batch
-
-    # ===== TRAIN BLOCK =====
-
-    def train_dataloader(self) -> DataLoader:
-        dataloader, n_samples = create_dataloader(
-            self.config.train_data_path,
-            self.config.max_context,
-            self.config.random_context,
-            self.config.shuffle_data,
-            self.config.batch_size,
-            self.num_workers,
-        )
-        print(f"approximate number of steps for train is {ceil(n_samples / self.config.batch_size)}")
-        return dataloader
-
     def training_step(self, batch: PathContextBatch, batch_idx: int) -> Dict:
         # [seq length; batch size; vocab size]
         logits = self(batch.context, batch.contexts_per_label, batch.labels.shape[0], batch.labels)
         loss = self._calculate_loss(logits, batch.labels)
-
+        log = {"train/loss": loss}
         with torch.no_grad():
-            subtoken_statistic = SubtokenStatistic.calculate_statistic(
-                batch.labels, logits.argmax(-1), [self.vocab.label_to_id[t] for t in [SOS, EOS, PAD, UNK]]
-            )
+            statistic = SubtokenStatistic().calculate_statistic(batch.labels, logits.argmax(-1))
```

Comment:
Maybe it will be more correct  to change the function `calculate_statistic` signature from `labels, logits` to ones similar to `cross_entopy`, namely: `logits, labels`

Reply:
It doesn't correspond to Code2Class model. We can do it later, in another pr.